libvast/src/command.cpp

```diff
@@ -105,53 +107,61 @@ std::string command::full_name() {
+                                         const option_map& options,
                                          argument_iterator begin,
                                          argument_iterator end) {
   VAST_UNUSED(options, begin, end);
-  VAST_TRACE(VAST_ARG(options), VAST_ARG("args", begin, end));
+  VAST_TRACE(VAST_ARG(std::string{name_}), VAST_ARG("args", begin, end),
+             VAST_ARG(options));
   return proceed_ok;
 }
 
-int command::run_impl(caf::actor_system&, option_map& options,
+int command::run_impl(caf::actor_system&, const option_map& options,
                       argument_iterator begin, argument_iterator end) {
   VAST_UNUSED(options, begin, end);
-  VAST_TRACE(VAST_ARG(options), VAST_ARG("args", begin, end));
+  VAST_TRACE(VAST_ARG(std::string{name_}), VAST_ARG("args", begin, end),
+             VAST_ARG(options));
   usage();
   return EXIT_FAILURE;
 }
 
-std::tuple<caf::message, std::string, caf::message>
-command::separate_args(const caf::message& args) {
-  auto arg = [&](size_t i) -> const std::string& {
-    VAST_ASSERT(args.match_element<std::string>(i));
-    return args.get_as<std::string>(i);
-  };
-  size_t pos = 0;
-  while (pos < args.size()) {
-    if (arg(pos).compare(0, 2, "--") == 0) {
-      // Simply skip over long options.
-      ++pos;
-    } else if (arg(pos).compare(0, 1, "-") == 0) {
-      // We assume short options always have an argument.
-      // TODO: we could look into the argument instead of just assuming it
-      //       always take an argument.
-      pos += 2;
-    } else {
-      // Found the end of the options list.
-      return std::make_tuple(args.take(pos), arg(pos), args.drop(pos+ 1));
-    }
-  }
-  return std::make_tuple(args, "", caf::none);
+expected<void> command::add_opt(std::string_view name,
+                                std::string_view description,
+                                data default_value) {
+  return opts_.add(name, description, std::move(default_value));
+}
+
+std::string command::parse_error(option_declaration_set::parse_state state,
+                                 argument_iterator error_position,
+                                 argument_iterator begin,
+                                 argument_iterator end) const {
+  std::stringstream result;
+  result << "Failed to parse command" << "\n";
+  result << "  Command: " << name() << " " << detail::join(begin, end, " ")
+         << "\n";
+  result << "  Description: " << to_string(state) << "\n";
+  if (error_position != end)
+    result << "  Position: " << *error_position << "\n";
+  else
+    result << "  Position: at the end\n";
+  result << "\n" << usage();
+  return result.str();
+}
+
+std::string command::subcommand_error(argument_iterator error_position,
+                                      argument_iterator,
+                                      argument_iterator end) const {
+  std::stringstream result;
+  result << "No such command: " << full_name();
+  if (error_position != end)
```

Comment:
Say I invoke `export`, then this function prints `No such command: export\n\n<help text>`.

To me that's a bit confusing, because `export` *is* a command. The problem is that no subcommand was specified.

Reply:
ok, I see the `subcommand_error()` is a misleading function name.  It is called when the sub-command `export` is not found.  In all other cases `parse_error()` is called. I can rename it to `subcommand_not_found()`.

Reply:
If you rename this to `subcommand_not_found` than `if (error_position != end)` should be an `ASSERT` instead.

Reply:
In this case the following typo in command `./vast imbort pcap < ...` would cause an abort of VAST because the position of the parser is at `imbort`which is not valid command.